corehq/apps/cleanup/management/commands/local_commcare_export.py

```diff
@@ -0,0 +1,231 @@
+from corehq.apps.api.resources.v0_4 import CommCareCaseResource, XFormInstanceResource
+from corehq.apps.api.serializers import CommCareCaseSerializer, XFormInstanceSerializer
+from corehq.apps.cloudcare.api import ElasticCaseQuery
+
+
+class MockApi(namedtuple('MockApi', 'query_set resource serializer')):
+
+    def serialize(self, obj):
+        return json.loads(self.serializer.serialize(self.resource.full_dehydrate(Bundle(obj=obj))))
+
+
+def _get_case_mock(project, params):
+    # this is mostly copy/paste/modified from CommCareCaseResource
+    filters = CaseListFilters(params).filters
+    query = ElasticCaseQuery(project, filters).get_query()
+    if 'from' in query:
+        del query['from']
+    if 'size' in query:
+        del query['size']
+
+    query_set = ElasticAPIQuerySet(
+        payload=query,
+        model=ESCase,
+        es_client=CaseES(project),
+    ).order_by('server_modified_on')
+
+    return MockApi(
+        query_set, CommCareCaseResource(), CommCareCaseSerializer()
+    )
+
+
+def _get_form_mock(project, params):
+    # this is mostly copy/paste/modified from XFormInstanceResource
+    include_archived = 'include_archived' in params
+    es_query = es_search_by_params(params, project, ['include_archived'])
+    if include_archived:
+        es_query['filter']['and'].append({'or': [
+            {'term': {'doc_type': 'xforminstance'}},
+            {'term': {'doc_type': 'xformarchived'}},
+        ]})
+    else:
+        es_query['filter']['and'].append({'term': {'doc_type': 'xforminstance'}})
+
+    query_set = ElasticAPIQuerySet(
+        payload=es_query,
+        model=ESXFormInstance,
+        es_client=XFormES(project),
+    ).order_by('-received_on')
+    return MockApi(
+        query_set, XFormInstanceResource(), XFormInstanceSerializer()
+    )
+
+
+def _get_mock_api(resource, project, params):
+    if resource == 'case':
+        return _get_case_mock(project, params)
+    elif resource == 'form':
+        return _get_form_mock(project, params)
+    else:
+        raise ValueError("Unknown/unsupported resource type '{}'".format(resource))
+
+
+class LocalCommCareHqClient(object):
+    """
+    Like CommCareHqClient but for a local environment
+    """
+    def __init__(self, url, project, checkpoint_manager=None):
+        self.url = url
+        self.project = project
+        self._checkpoint_manager = checkpoint_manager
+
+    # todo: backoff?
+    # @backoff.on_exception(
+    #     backoff.expo, requests.exceptions.RequestException,
+    #     max_time=300, giveup=is_client_error,
+    #     on_backoff=on_backoff, on_giveup=on_giveup
+    # )
+    def get(self, es_query_set, start, limit, params=None):
+        """
+        """
+        from commcare_export.cli import logger
+        logger.debug("Fetching batch: {}-{}".format(start, limit))
+        # logger.debug("Fetching batch: %s", params)
+        return es_query_set[start:start + limit]
```

Comment:
This will slow down once the pagination starts to get deeper. Might want to consider using date based pagination.

Reply:
I looked into this and since this is managed by the `env` it's kind of tricky to do cleanly. I think we could do something like paginate through some fixed multiple of `limit` and then break all the way out, re-update the variable in the `env` based on the checkpoint and continue, but it gets pretty hairy. The other option would be to edit more code in `commcare-export` but I was kind of hoping to avoid that.

FYI, after ~24 hours and 1.75 million forms the rate of processing had dropped from about 1500 forms/minute to 1200 forms/minute. So it does slow down but it's not too bad. Also you can just restart the process and it resumes at 0 from the checkpoint.

Reply:
(it's also possible i'm missing an obvious/dumb way to do it in which case i'd love to be told that)